File: python/matching-brackets/matching_brackets.py
def is_paired(input_string):
    stack = []
    for ch in filter(lambda ch: ch in '()[]{}', input_string):
        if ch in '([{':
            stack.append(ch)
        elif len(stack) == 0 or (ch == ')' and stack[-1] != '(') or (ch == ']' and stack[-1] != '[') or (ch == '}' and stack[-1] != '{'):
            return False
        else:
            stack.pop()
    return len(stack) == 0

File: cpp/matching-brackets/matching_brackets.cpp
#include "matching_brackets.h"
#include <stack>

namespace matching_brackets
{
bool check(const std::string &s)
{
    std::stack<char> stack;
    for (char ch : s)
    {
        if (ch == '(' || ch == '[' || ch == '{')
        {
            stack.push(ch);
        }
        else if (ch == ')' || ch == ']' || ch == '}')
        {
            if (stack.empty() ||
                (ch == ')' && stack.top() != '(') ||
                (ch == ']' && stack.top() != '[') ||
                (ch == '}' && stack.top() != '{'))
            {
                return false;
            }
            else
            {
                stack.pop();
            }
        }
    }

    return stack.empty();
}
} // namespace matching_brackets
